src/api/crypto/Hash.cpp

```diff
@@ -90,6 +98,127 @@ auto Hash::bitcoin_hash_160(
     }
 }
 
+auto Hash::ethereum_address(
+    const ReadView keccakData,
+    const AllocateOutput destination) const -> bool
+{
+    try {
+        const int eth_address_length = 20;
+        auto keccak_data_bytes =
+            reinterpret_cast<const uint8_t*>(keccakData.data());
+        const uint8_t* last_20_data_bytes =
+            keccak_data_bytes + keccakData.size() - eth_address_length;
+
+        auto address = destination(eth_address_length);
+
+        if (false == address.valid(eth_address_length)) {
+            LogError()(__func__)(": Failed to allocate space for an address")
+                .Flush();
+
+            return false;
+        }
+
+        address.as<unsigned char>();
+        std::memcpy(address, last_20_data_bytes, eth_address_length);
+
+        return true;
+    } catch (const std::exception& e) {
+        LogError()(OT_PRETTY_CLASS())(e.what()).Flush();
+
+        return false;
+    }
+}
+
+auto Hash::ethereum_hash(const ReadView data, const AllocateOutput destination)
+    const noexcept -> bool
+{
+    try {
+        auto uncompressed_pubkey_temp = Space{};
+        const auto success =
+            uncompress_pubkey(data, writer(uncompressed_pubkey_temp));
+        if (false == success) {
+
+            throw std::runtime_error{"failed to calculate uncompressed pubkey"};
+        }
+        ReadView uncompressed_pubkey_without_prefix_readview(
+            reinterpret_cast<const char*>(uncompressed_pubkey_temp.data()),
+            uncompressed_pubkey_temp.size());
+
+        auto keccak_temp =
+            space(Provider::HashSize(opentxs::crypto::HashType::Keccak256));
+
+        const auto rc = Digest(
+            opentxs::crypto::HashType::Keccak256,
+            uncompressed_pubkey_without_prefix_readview,
+            writer(keccak_temp));
+
+        if (false == rc) {
+
+            throw std::runtime_error{
+                "failed to calculate intermediate keccak hash"};
+        }
+
+        return ethereum_address(reader(keccak_temp), destination);
+    } catch (const std::exception& e) {
+        LogError()(OT_PRETTY_CLASS())(e.what()).Flush();
+
+        return false;
+    }
+}
+
+auto Hash::uncompress_pubkey(const ReadView data, const AllocateOutput output)
+    const noexcept -> bool
+{
+    try {
+        if (false == output.operator bool()) {
+            throw std::runtime_error{"invalid output"};
+        }
+
+        if (data.size() > std::numeric_limits<int>::max()) {
+            throw std::runtime_error{"input too large"};
+        }
```

Comment:
does it prevent anything? shouldn't we just check if data > expected compressed pubkey size? (33 bytes or sth along this lines)

Reply:
I took this code from 
```
auto OpenSSL::Digest(
    const crypto::HashType type,
    const ReadView data,
    const AllocateOutput output) const noexcept -> bool
```
but it doesn't do anything helpful

Reply:
it makes sense in 'Digest' as it just hashes any arbitrary data (with the limit on max_size).  
here, on the other hand, we have strict indication that we want to uncompress a key and we know the expected keysize in advance so we can use it to verify that ;)